File: TouchNothing/Utilities/AuthResponseDecoder.swift
import Foundation

enum AuthResponseDecoder {
    static func decodeSession(from data: Data, fallbackNickname: String) throws -> AuthSession {
        if let payload = try? JSONDecoder().decode(AuthRPCResponse.self, from: data) {
            if let serviceError = payload.serviceError {
                throw serviceError
            }
            if let session = payload.authSession {
                return session
            }
        }

        if let session = try? JSONDecoder().decode(AuthSession.self, from: data) {
            return session
        }

        if let token = try? JSONDecoder().decode(String.self, from: data) {
            return AuthSession(
                nickname: NicknameValidator.normalized(fallbackNickname),
                sessionToken: token
            )
        }

        throw SupabaseServiceError.serverUpgradeRequired
    }
}

File: app/src/main/java/com/touchnothing/app/data/remote/AuthResponseDecoder.kt
package com.touchnothing.app.data.remote

import com.touchnothing.app.data.model.AuthRpcResponse
import com.touchnothing.app.data.model.AuthSession
import com.touchnothing.app.util.NicknameValidator
import kotlinx.serialization.json.Json

object AuthResponseDecoder {
    private val json = Json { ignoreUnknownKeys = true }

    fun decodeSession(data: String, fallbackNickname: String): AuthSession {
        runCatching {
            val payload = json.decodeFromString<AuthRpcResponse>(data)
            payload.serviceError?.let { throw it }
            payload.authSession?.let { return it }
        }

        runCatching {
            return json.decodeFromString<AuthSession>(data)
        }

        runCatching {
            val token = json.decodeFromString<String>(data)
            return AuthSession(
                nickname = NicknameValidator.normalized(fallbackNickname),
                sessionToken = token,
            )
        }

        throw SupabaseServiceError.ServerUpgradeRequired
    }
}

private val AuthRpcResponse.authSession: AuthSession?
    get() {
        if (ok == false) return null
        val nick = nickname ?: return null
        val token = sessionToken ?: return null
        return AuthSession(nickname = nick, sessionToken = token)
    }

private val AuthRpcResponse.serviceError: SupabaseServiceError?
    get() {
        if (ok != false) return null
        if (error == "invalid_credentials") {
            return SupabaseServiceError.InvalidCredentials(remainingAttempts)
        }
        return SupabaseRpcErrorMapper.mapAuthErrorToken(error) ?: SupabaseServiceError.NetworkFailure
    }
